relayer/src/chain/cosmos/encode.rs

```diff
@@ -0,0 +1,179 @@
+use bech32::{ToBase32, Variant};
+use core::str::FromStr;
+use ibc::core::ics24_host::identifier::ChainId;
+use ibc_proto::cosmos::tx::v1beta1::mode_info::{Single, Sum};
+use ibc_proto::cosmos::tx::v1beta1::{AuthInfo, Fee, ModeInfo, SignDoc, SignerInfo, TxBody, TxRaw};
+use ibc_proto::google::protobuf::Any;
+use tendermint::account::Id as AccountId;
+
+use crate::chain::cosmos::types::account::{Account, AccountNumber, AccountSequence};
+use crate::chain::cosmos::types::tx::SignedTx;
+use crate::config::types::Memo;
+use crate::config::AddressType;
+use crate::config::ChainConfig;
+use crate::error::Error;
+use crate::keyring::{sign_message, KeyEntry};
+
+pub fn sign_and_encode_tx(
+    config: &ChainConfig,
+    key_entry: &KeyEntry,
+    account: &Account,
+    tx_memo: &Memo,
+    messages: Vec<Any>,
+    fee: &Fee,
+) -> Result<Vec<u8>, Error> {
+    let signed_tx = sign_tx(config, key_entry, account, tx_memo, messages, fee)?;
+
+    let tx_raw = TxRaw {
+        body_bytes: signed_tx.body_bytes,
+        auth_info_bytes: signed_tx.auth_info_bytes,
+        signatures: signed_tx.signatures,
+    };
+
+    encode_tx_raw(tx_raw)
+}
+
+pub fn sign_tx(
+    config: &ChainConfig,
+    key_entry: &KeyEntry,
+    account: &Account,
+    tx_memo: &Memo,
+    messages: Vec<Any>,
+    fee: &Fee,
+) -> Result<SignedTx, Error> {
+    let key_bytes = encode_key_bytes(key_entry)?;
+
+    let signer = encode_signer_info(&config.address_type, account.sequence, key_bytes)?;
+
+    let (body, body_bytes) = tx_body_and_bytes(messages, tx_memo)?;
+
+    let (auth_info, auth_info_bytes) = auth_info_and_bytes(signer, fee.clone())?;
+
+    let signed_doc = encode_sign_doc(
+        &config.id,
+        key_entry,
+        &config.address_type,
+        account.number,
+        auth_info_bytes.clone(),
+        body_bytes.clone(),
+    )?;
+
+    Ok(SignedTx {
+        body,
+        body_bytes,
+        auth_info,
+        auth_info_bytes,
+        signatures: vec![signed_doc],
+    })
+}
+
+fn encode_key_bytes(key: &KeyEntry) -> Result<Vec<u8>, Error> {
```

Comment:
This is refactored from:

https://github.com/informalsystems/ibc-rs/blob/3e34b4b916eb91f680412e5fdd3ea045a99526d2/relayer/src/chain/cosmos.rs#L690-L695